File: eve_mcp/cache.py
"""SQLite-backed HTTP cache and id->name store.

ESI bans clients that circumvent its caching, so every GET goes through here:
inside the ``expires`` window we never touch the network at all, and afterwards
we revalidate with ``If-None-Match`` instead of refetching the body.
"""
from __future__ import annotations

import asyncio
import json
import sqlite3
import time
from dataclasses import dataclass
from pathlib import Path
from typing import Any

_SCHEMA = """
CREATE TABLE IF NOT EXISTS http_cache (
    key        TEXT PRIMARY KEY,
    etag       TEXT,
    expires_at REAL NOT NULL,
    stored_at  REAL NOT NULL,
    pages      INTEGER,
    body       TEXT NOT NULL
);
CREATE INDEX IF NOT EXISTS http_cache_expires ON http_cache (expires_at);

CREATE TABLE IF NOT EXISTS names (
    id       INTEGER PRIMARY KEY,
    name     TEXT NOT NULL,
    category TEXT
);

CREATE TABLE IF NOT EXISTS blobs (
    key       TEXT PRIMARY KEY,
    stored_at REAL NOT NULL,
    value     TEXT NOT NULL
);
"""


@dataclass
class CachedResponse:
    body: Any
    etag: str | None
    expires_at: float
    stored_at: float
    pages: int | None

    @property
    def fresh(self) -> bool:
        return time.time() < self.expires_at

    @property
    def age_seconds(self) -> float:
        return max(0.0, time.time() - self.stored_at)


class Store:
    """Thin async wrapper over a single sqlite connection."""

    def __init__(self, path: Path) -> None:
        path.parent.mkdir(parents=True, exist_ok=True)
        self._conn = sqlite3.connect(path, check_same_thread=False)
        self._conn.row_factory = sqlite3.Row
        self._conn.execute("PRAGMA journal_mode=WAL")
        self._conn.execute("PRAGMA synchronous=NORMAL")
        self._conn.executescript(_SCHEMA)
        self._conn.commit()
        self._lock = asyncio.Lock()

    async def close(self) -> None:
        async with self._lock:
            self._conn.close()

    # ------------------------------------------------------------------ http

    async def get_http(self, key: str) -> CachedResponse | None:
        async with self._lock:
            row = self._conn.execute(
                "SELECT etag, expires_at, stored_at, pages, body FROM http_cache WHERE key = ?",
                (key,),
            ).fetchone()
        if row is None:
            return None
        try:
            body = json.loads(row["body"])
        except json.JSONDecodeError:
            return None
        return CachedResponse(
            body=body,
            etag=row["etag"],
            expires_at=row["expires_at"],
            stored_at=row["stored_at"],
            pages=row["pages"],
        )

    async def put_http(
        self,
        key: str,
        body: Any,
        etag: str | None,
        expires_at: float,
        pages: int | None = None,
    ) -> None:
        now = time.time()
        async with self._lock:
            self._conn.execute(
                "INSERT INTO http_cache (key, etag, expires_at, stored_at, pages, body) "
                "VALUES (?, ?, ?, ?, ?, ?) "
                "ON CONFLICT(key) DO UPDATE SET etag=excluded.etag, "
                "expires_at=excluded.expires_at, stored_at=excluded.stored_at, "
                "pages=excluded.pages, body=excluded.body",
                (key, etag, expires_at, now, pages, json.dumps(body)),
            )
            self._conn.commit()

    async def touch_http(self, key: str, expires_at: float) -> None:
        """Extend the freshness window after a 304 Not Modified."""
        async with self._lock:
            self._conn.execute(
                "UPDATE http_cache SET expires_at = ?, stored_at = ? WHERE key = ?",
                (expires_at, time.time(), key),
            )
            self._conn.commit()

    async def purge_expired(self, older_than_days: float = 30.0) -> int:
        cutoff = time.time() - older_than_days * 86400
        async with self._lock:
            cur = self._conn.execute("DELETE FROM http_cache WHERE stored_at < ?", (cutoff,))
            self._conn.commit()
            return cur.rowcount

    # ----------------------------------------------------------------- names

    async def get_names(self, ids: list[int]) -> dict[int, dict[str, Any]]:
        if not ids:
            return {}
        out: dict[int, dict[str, Any]] = {}
        async with self._lock:
            for chunk_start in range(0, len(ids), 500):
                chunk = ids[chunk_start : chunk_start + 500]
                placeholders = ",".join("?" * len(chunk))
                rows = self._conn.execute(
                    f"SELECT id, name, category FROM names WHERE id IN ({placeholders})",
                    chunk,
                ).fetchall()
                for row in rows:
                    out[row["id"]] = {"name": row["name"], "category": row["category"]}
        return out

    async def put_names(self, entries: list[tuple[int, str, str | None]]) -> None:
        if not entries:
            return
        async with self._lock:
            self._conn.executemany(
                "INSERT INTO names (id, name, category) VALUES (?, ?, ?) "
                "ON CONFLICT(id) DO UPDATE SET name=excluded.name, category=excluded.category",
                entries,
            )
            self._conn.commit()

    # ----------------------------------------------------------------- blobs

    async def get_blob(self, key: str, max_age: float | None = None) -> Any | None:
        async with self._lock:
            row = self._conn.execute(
                "SELECT stored_at, value FROM blobs WHERE key = ?", (key,)
            ).fetchone()
        if row is None:
            return None
        if max_age is not None and time.time() - row["stored_at"] > max_age:
            return None
        try:
            return json.loads(row["value"])
        except json.JSONDecodeError:
            return None

    async def put_blob(self, key: str, value: Any) -> None:
        async with self._lock:
            self._conn.execute(
                "INSERT INTO blobs (key, stored_at, value) VALUES (?, ?, ?) "
                "ON CONFLICT(key) DO UPDATE SET stored_at=excluded.stored_at, value=excluded.value",
                (key, time.time(), json.dumps(value)),
            )
            self._conn.commit()

File: internal/adapter/cache/cache.go
package cache

import (
	"database/sql"
	"encoding/json"
	"os"
	"path/filepath"
	"sync"
	"time"

	_ "modernc.org/sqlite"
)

type CachedResponse struct {
	Body      any
	ETag      string
	ExpiresAt float64
	StoredAt  float64
	Pages     *int
}

func (c *CachedResponse) Fresh() bool {
	return time.Now().Unix() < int64(c.ExpiresAt)
}

func (c *CachedResponse) AgeSeconds() float64 {
	age := time.Now().Sub(time.Unix(int64(c.StoredAt), 0)).Seconds()
	if age < 0 {
		return 0
	}
	return age
}

type Store struct {
	db *sql.DB
	mu sync.Mutex
}

func Open(path string) (*Store, error) {
	if err := os.MkdirAll(filepath.Dir(path), 0o700); err != nil {
		return nil, err
	}
	db, err := sql.Open("sqlite", path)
	if err != nil {
		return nil, err
	}
	if _, err := db.Exec(`PRAGMA journal_mode=WAL`); err != nil {
		db.Close()
		return nil, err
	}
	if _, err := db.Exec(`PRAGMA synchronous=NORMAL`); err != nil {
		db.Close()
		return nil, err
	}
	if _, err := db.Exec(`
CREATE TABLE IF NOT EXISTS http_cache (
    key        TEXT PRIMARY KEY,
    etag       TEXT,
    expires_at REAL NOT NULL,
    stored_at  REAL NOT NULL,
    pages      INTEGER,
    body       TEXT NOT NULL
);
CREATE INDEX IF NOT EXISTS http_cache_expires ON http_cache (expires_at);
CREATE TABLE IF NOT EXISTS names (
    id       INTEGER PRIMARY KEY,
    name     TEXT NOT NULL,
    category TEXT
);
CREATE TABLE IF NOT EXISTS blobs (
    key       TEXT PRIMARY KEY,
    stored_at REAL NOT NULL,
    value     TEXT NOT NULL
);`); err != nil {
		db.Close()
		return nil, err
	}
	return &Store{db: db}, nil
}

func (s *Store) Close() error { return s.db.Close() }

func (s *Store) GetHTTP(key string) (*CachedResponse, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	var etag sql.NullString
	var expires, stored float64
	var pages sql.NullInt64
	var body string
	err := s.db.QueryRow(
		`SELECT etag, expires_at, stored_at, pages, body FROM http_cache WHERE key = ?`, key,
	).Scan(&etag, &expires, &stored, &pages, &body)
	if err == sql.ErrNoRows {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	var decoded any
	if err := json.Unmarshal([]byte(body), &decoded); err != nil {
		return nil, nil
	}
	c := &CachedResponse{Body: decoded, ExpiresAt: expires, StoredAt: stored}
	if etag.Valid {
		c.ETag = etag.String
	}
	if pages.Valid {
		p := int(pages.Int64)
		c.Pages = &p
	}
	return c, nil
}

func (s *Store) PutHTTP(key string, body any, etag string, expiresAt float64, pages *int) error {
	raw, err := json.Marshal(body)
	if err != nil {
		return err
	}
	var pagesVal any
	if pages != nil {
		pagesVal = *pages
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	_, err = s.db.Exec(
		`INSERT INTO http_cache (key, etag, expires_at, stored_at, pages, body)
		 VALUES (?, ?, ?, ?, ?, ?)
		 ON CONFLICT(key) DO UPDATE SET etag=excluded.etag, expires_at=excluded.expires_at,
		 stored_at=excluded.stored_at, pages=excluded.pages, body=excluded.body`,
		key, nullIfEmpty(etag), expiresAt, float64(time.Now().Unix()), pagesVal, string(raw),
	)
	return err
}

func (s *Store) TouchHTTP(key string, expiresAt float64) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	_, err := s.db.Exec(
		`UPDATE http_cache SET expires_at = ?, stored_at = ? WHERE key = ?`,
		expiresAt, float64(time.Now().Unix()), key,
	)
	return err
}

func (s *Store) PurgeExpired(olderThanDays float64) (int64, error) {
	cutoff := float64(time.Now().Unix()) - olderThanDays*86400
	s.mu.Lock()
	defer s.mu.Unlock()
	res, err := s.db.Exec(`DELETE FROM http_cache WHERE stored_at < ?`, cutoff)
	if err != nil {
		return 0, err
	}
	return res.RowsAffected()
}

func (s *Store) GetNames(ids []int) (map[int]NameRow, error) {
	out := map[int]NameRow{}
	if len(ids) == 0 {
		return out, nil
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	for start := 0; start < len(ids); start += 500 {
		end := start + 500
		if end > len(ids) {
			end = len(ids)
		}
		chunk := ids[start:end]
		query := `SELECT id, name, category FROM names WHERE id IN (` + placeholders(len(chunk)) + `)`
		args := make([]any, len(chunk))
		for i, id := range chunk {
			args[i] = id
		}
		rows, err := s.db.Query(query, args...)
		if err != nil {
			return nil, err
		}
		for rows.Next() {
			var id int
			var name string
			var cat sql.NullString
			if err := rows.Scan(&id, &name, &cat); err != nil {
				rows.Close()
				return nil, err
			}
			out[id] = NameRow{Name: name, Category: cat.String}
		}
		rows.Close()
	}
	return out, nil
}

type NameRow struct {
	Name     string
	Category string
}

func (s *Store) PutNames(entries [][3]any) error {
	if len(entries) == 0 {
		return nil
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	tx, err := s.db.Begin()
	if err != nil {
		return err
	}
	stmt, err := tx.Prepare(
		`INSERT INTO names (id, name, category) VALUES (?, ?, ?)
		 ON CONFLICT(id) DO UPDATE SET name=excluded.name, category=excluded.category`,
	)
	if err != nil {
		tx.Rollback()
		return err
	}
	for _, e := range entries {
		if _, err := stmt.Exec(e[0], e[1], e[2]); err != nil {
			stmt.Close()
			tx.Rollback()
			return err
		}
	}
	stmt.Close()
	return tx.Commit()
}

func (s *Store) GetBlob(key string, maxAge *float64) (any, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	var stored float64
	var value string
	err := s.db.QueryRow(`SELECT stored_at, value FROM blobs WHERE key = ?`, key).Scan(&stored, &value)
	if err == sql.ErrNoRows {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	if maxAge != nil && float64(time.Now().Unix())-stored > *maxAge {
		return nil, nil
	}
	var decoded any
	if err := json.Unmarshal([]byte(value), &decoded); err != nil {
		return nil, nil
	}
	return decoded, nil
}

func (s *Store) PutBlob(key string, value any) error {
	raw, err := json.Marshal(value)
	if err != nil {
		return err
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	_, err = s.db.Exec(
		`INSERT INTO blobs (key, stored_at, value) VALUES (?, ?, ?)
		 ON CONFLICT(key) DO UPDATE SET stored_at=excluded.stored_at, value=excluded.value`,
		key, float64(time.Now().Unix()), string(raw),
	)
	return err
}

func placeholders(n int) string {
	if n <= 0 {
		return ""
	}
	b := make([]byte, 0, n*2)
	for i := 0; i < n; i++ {
		if i > 0 {
			b = append(b, ',')
		}
		b = append(b, '?')
	}
	return string(b)
}

func nullIfEmpty(s string) any {
	if s == "" {
		return nil
	}
	return s
}
